<div align="center"> <h1>Imputation of Missing Values</h1>
    <h2><a href="...">Richard Leibrandt</a></h2>
</div>

Sometimes datasets are incomplete. For example, the column engine-location has missing values in the following:

In [1]:
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import StandardScaler
from sklearn.impute import SimpleImputer
from sklearn.linear_model import LinearRegression
import pandas as pd

car_prices=pd.read_csv("../data/car_prices.csv")

features = car_prices
features = features.replace('?', '', regex=False)
features = features.apply(pd.to_numeric, errors='coerce')

In [2]:
features.isna().sum()

symboling              0
normalized-losses     41
make                 205
fuel-type            205
aspiration           205
num-of-doors         205
body-style           205
drive-wheels         205
engine-location      205
wheel-base             0
length                 0
width                  0
height                 0
curb-weight            0
engine-type          205
num-of-cylinders     205
engine-size            0
fuel-system          205
bore                   4
stroke                 4
compression-ratio      0
horsepower             2
peak-rpm               2
city-mpg               0
highway-mpg            0
price                  4
dtype: int64

First, let's remove all columns that are compeltely empty:

In [3]:
features = features.dropna(axis='columns', how='all')

In [4]:
features.describe()

symboling  normalized-losses  wheel-base      length       width  \
count  205.000000         164.000000  205.000000  205.000000  205.000000   
mean     0.834146         122.000000   98.756585  174.049268   65.907805   
std      1.245307          35.442168    6.021776   12.337289    2.145204   
min     -2.000000          65.000000   86.600000  141.100000   60.300000   
25%      0.000000          94.000000   94.500000  166.300000   64.100000   
50%      1.000000         115.000000   97.000000  173.200000   65.500000   
75%      2.000000         150.000000  102.400000  183.100000   66.900000   
max      3.000000         256.000000  120.900000  208.100000   72.300000   

           height  curb-weight  engine-size        bore      stroke  \
count  205.000000   205.000000   205.000000  201.000000  201.000000   
mean    53.724878  2555.565854   126.907317    3.329751    3.255423   
std      2.443522   520.680204    41.642693    0.273539    0.316717   
min     47.800000  1488.000000    61.000000    2.540000    2.070000   
25%     52.000000  2145.000000    97.000000    3.150000    3.110000   
50%     54.100000  2414.000000   120.000000    3.310000    3.290000   
75%     55.500000  2935.000000   141.000000    3.590000    3.410000   
max     59.800000  4066.000000   326.000000    3.940000    4.170000   

       compression-ratio  horsepower     peak-rpm    city-mpg  highway-mpg  \
count         205.000000  203.000000   203.000000  205.000000   205.000000   
mean           10.142537  104.256158  5125.369458   25.219512    30.751220   
std             3.972040   39.714369   479.334560    6.542142     6.886443   
min             7.000000   48.000000  4150.000000   13.000000    16.000000   
25%             8.600000   70.000000  4800.000000   19.000000    25.000000   
50%             9.000000   95.000000  5200.000000   24.000000    30.000000   
75%             9.400000  116.000000  5500.000000   30.000000    34.000000   
max            23.000000  288.000000  6600.000000   49.000000    54.000000   

              price  
count    201.000000  
mean   13207.129353  
std     7947.066342  
min     5118.000000  
25%     7775.000000  
50%    10295.000000  
75%    16500.000000  
max    45400.000000

Most algorithms of machine learning and data mining do not accept inputs with missing values. There are several strategies to deal with missing values:

* remove observations that contain missing values
* imputation of missing values

Removing observations with missing values is easy with pandas:

In [5]:
features.dropna(axis='index', how='any').describe()

symboling  normalized-losses  wheel-base      length       width  \
count  160.000000         160.000000  160.000000  160.000000  160.000000   
mean     0.737500         121.300000   98.235625  172.319375   65.596250   
std      1.189511          35.602417    5.163763   11.548860    1.946999   
min     -2.000000          65.000000   86.600000  141.100000   60.300000   
25%      0.000000          94.000000   94.500000  165.525000   64.000000   
50%      1.000000         114.000000   96.900000  172.200000   65.400000   
75%      2.000000         148.000000  100.600000  177.800000   66.500000   
max      3.000000         256.000000  115.600000  202.600000   71.700000   

           height  curb-weight  engine-size        bore      stroke  \
count  160.000000   160.000000   160.000000  160.000000  160.000000   
mean    53.878750  2459.450000   119.093750    3.298437    3.237313   
std      2.276608   480.897834    30.411186    0.267348    0.294210   
min     49.400000  1488.000000    61.000000    2.540000    2.070000   
25%     52.000000  2073.250000    97.000000    3.050000    3.107500   
50%     54.100000  2338.500000   110.000000    3.270000    3.270000   
75%     55.500000  2808.750000   134.500000    3.550000    3.410000   
max     59.800000  4066.000000   258.000000    3.940000    4.170000   

       compression-ratio  horsepower     peak-rpm    city-mpg  highway-mpg  \
count         160.000000  160.000000   160.000000  160.000000   160.000000   
mean           10.145125   95.875000  5116.250000   26.506250    32.068750   
std             3.882507   30.625708   465.290536    6.081208     6.440948   
min             7.000000   48.000000  4150.000000   15.000000    18.000000   
25%             8.700000   69.000000  4800.000000   23.000000    28.000000   
50%             9.000000   88.000000  5200.000000   26.000000    32.000000   
75%             9.400000  114.000000  5500.000000   31.000000    37.000000   
max            23.000000  200.000000  6600.000000   49.000000    54.000000   

              price  
count    160.000000  
mean   11427.681250  
std     5863.789011  
min     5118.000000  
25%     7383.500000  
50%     9164.000000  
75%    14559.250000  
max    35056.000000

We can also consider only specific columns for the drop:

In [6]:
features.dropna(axis=0, subset=['price']).head()

symboling  normalized-losses  wheel-base  length  width  height  \
0          3                NaN        88.6   168.8   64.1    48.8   
1          3                NaN        88.6   168.8   64.1    48.8   
2          1                NaN        94.5   171.2   65.5    52.4   
3          2              164.0        99.8   176.6   66.2    54.3   
4          2              164.0        99.4   176.6   66.4    54.3   

   curb-weight  engine-size  bore  stroke  compression-ratio  horsepower  \
0         2548          130  3.47    2.68                9.0       111.0   
1         2548          130  3.47    2.68                9.0       111.0   
2         2823          152  2.68    3.47                9.0       154.0   
3         2337          109  3.19    3.40               10.0       102.0   
4         2824          136  3.19    3.40                8.0       115.0   

   peak-rpm  city-mpg  highway-mpg    price  
0    5000.0        21           27  13495.0  
1    5000.0        21           27  16500.0  
2    5000.0        19           26  16500.0  
3    5500.0        24           30  13950.0  
4    5500.0        18           22  17450.0

With this solution we keep only 160 observations. We lost more than 20% of the observations.

Let's try the other approach. We can impute the missing values, inferring them from the rest of the known part of the data. Below, the missing values are replaced by the mean of the known values. The other available strategies are the median and the most frequent value.

Note that observations with missing values in the target feature are useless to 'teach' the model, so we get rid of them first.

In [7]:
from sklearn.impute import SimpleImputer

imputer = SimpleImputer(strategy='mean')  # define the imputer object
features_imp = imputer.fit_transform(features)  # fit to the data and tranform it
features_imp.shape

(205, 16)

The output is a numpy array. For an easier visualization we can transform it back to a pandas Dataframe:

In [8]:
features_df=pd.DataFrame(features_imp, columns=features.columns)
features_df.head()

symboling  normalized-losses  wheel-base  length  width  height  \
0        3.0              122.0        88.6   168.8   64.1    48.8   
1        3.0              122.0        88.6   168.8   64.1    48.8   
2        1.0              122.0        94.5   171.2   65.5    52.4   
3        2.0              164.0        99.8   176.6   66.2    54.3   
4        2.0              164.0        99.4   176.6   66.4    54.3   

   curb-weight  engine-size  bore  stroke  compression-ratio  horsepower  \
0       2548.0        130.0  3.47    2.68                9.0       111.0   
1       2548.0        130.0  3.47    2.68                9.0       111.0   
2       2823.0        152.0  2.68    3.47                9.0       154.0   
3       2337.0        109.0  3.19    3.40               10.0       102.0   
4       2824.0        136.0  3.19    3.40                8.0       115.0   

   peak-rpm  city-mpg  highway-mpg    price  
0    5000.0      21.0         27.0  13495.0  
1    5000.0      21.0         27.0  16500.0  
2    5000.0      19.0         26.0  16500.0  
3    5500.0      24.0         30.0  13950.0  
4    5500.0      18.0         22.0  17450.0